src/main/java/no/ks/fiks/io/asice/read/EncryptedAsicReaderImpl.java

```diff
@@ -65,25 +67,29 @@ public ZipInputStream decrypt(final InputStream encryptedAsicData, final Private
     }
 
     @Override
-    public void writeDecryptedToPath(InputStream encryptedAsicData, PrivateKey privateKey, Path targetPath) {
+    public void writeDecryptedToPath(InputStream encryptedAsicData, List<PrivateKey> privateKeys, Path targetPath) {
         Preconditions.checkNotNull(encryptedAsicData);
-        Preconditions.checkNotNull(privateKey);
+        Preconditions.checkNotNull(privateKeys);
         Preconditions.checkNotNull(targetPath);
         try (OutputStream fileStream = Files.newOutputStream(targetPath);
              ZipOutputStream zipOutputStream = new ZipOutputStream(fileStream)) {
-            decrypt(encryptedAsicData, privateKey, zipOutputStream);
+            decrypt(encryptedAsicData, privateKeys, zipOutputStream);
         } catch (Exception e) {
             throw new RuntimeException(e);
         }
     }
 
-    private void decrypt(final InputStream encryptedAsic, final PrivateKey privateKey, final ZipOutputStream zipOutputStream) {
-        CMSKrypteringImpl cmsKryptering = new CMSKrypteringImpl();
-
-        InputStream inputStream = cmsKryptering.dekrypterData(encryptedAsic, privateKey);
-        decryptElementer(encryptedAsic, zipOutputStream, inputStream);
+    private void decrypt(final InputStream encryptedAsic, final List<PrivateKey> privateKeys, final ZipOutputStream zipOutputStream) {
+        if(!encryptedAsic.markSupported()) {
+            InputStream inputStream = cmsKrypteringHandler.handleEncryptedStream(new BufferedInputStream(encryptedAsic), privateKeys);
+            decryptElementer(encryptedAsic, zipOutputStream, inputStream);
```

Comment:
Man går fra å instanstiere en ny CMSKryperingImpl hver gang til å gjenbruke samme. Er det threadsafe?

Reply:
Jeg merket meg at for DecryptionStreamService (decrypt) ble samme CMSKrypteringImpl gjenbrukt mens for EncryptedSasicReader (writeDecrypted) ble det instansiert ny for hver gang. Å bruke samme virker for meg mer hensiktsmessig, og jeg ser ikke noen egenskapsmessig forskjell mellom de to som tilsier at det bør gjøres forskjellig, men kanskje? StreamService bruker en executorService, asicReader gjør ikke...

Reply:
Bør sjekkes ut hvorfor det ble gjort som det ble gjort i gammel versjon.